### PR TITLE
Add support for Ravnur Media Services

### PR DESCRIPTION
**Pull request by vstratovych**
## Purpose
This PR adds possibility to connect to RMS instance Ravnur Media Services is one of the possible AMS replacements.
Thanks to identical public API it's possible to use the same SDK and easily switch existing application from AMS to RMS.

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[x] User Interface
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What to Check
Verify that the following are valid
* Smoke test that all basic features are working

## Other Information
Detailed documentation about Ravnur Media Services can be found here:
https://github.com/Ravnur-Inc/ams-api-replacement-demo-app

### CODE REVIEW

File: AMSExplorer/AMSClient/AMSClientV3.cs [hunk 1]
```diff
@@ -28,6 +23,16 @@
...
+using AMSClient;
+
+using AMSExplorer.Ravnur;
+
+using Azure.ResourceManager;
+using Azure.ResourceManager.Media;
+
+using Microsoft.Identity.Client;
+using Microsoft.Rest;
+
```
Comment (dima-ravnur) [inline AMSExplorer/AMSClient/AMSClientV3.cs]: Lets don't touch things which aren't related to our change

File: AMSExplorer/DataGrid/DataGridViewStreamingEndpoints.cs [hunk 2]
```diff
@@ -88,7 +88,7 @@ public async Task InitAsync(AMSClientV3 amsClient)
                     Name = o.Data.Name,
                     Id = o.Data.Id,
                     Description = o.Data.Description,
-                    CDN = ((bool)o.Data.IsCdnEnabled) ? StreamingEndpointInformation.ReturnDisplayedProvider(o.Data.CdnProvider) ?? "CDN" : string.Empty,
+                    CDN = ((bool)o.Data.IsCdnEnabled.GetValueOrDefault()) ? StreamingEndpointInformation.ReturnDisplayedProvider(o.Data.CdnProvider) ?? "CDN" : string.Empty,
```
Comment (dima-ravnur) [inline AMSExplorer/DataGrid/DataGridViewStreamingEndpoints.cs]: If this field is missing on our side. I'd suggest to add it to our response and do not change this code

File: AMSExplorer/DataGrid/DataGridViewStreamingEndpoints.cs [hunk 3]
```diff
@@ -133,7 +133,7 @@ public async Task RefreshStreamingEndpointAsync(StreamingEndpointResource stream
                     _MyObservStreamingEndpoints[index].Description = streamingEndpoint.Data.Description;
                     _MyObservStreamingEndpoints[index].LastModifiedOn = streamingEndpoint.Data.LastModifiedOn?.DateTime.ToLocalTime();
                     _MyObservStreamingEndpoints[index].Type = StreamingEndpointInformation.ReturnTypeSE(streamingEndpoint);
-                    _MyObservStreamingEndpoints[index].CDN = ((bool)streamingEndpoint.Data.IsCdnEnabled) ? StreamingEndpointInformation.ReturnDisplayedProvider(streamingEndpoint.Data.CdnProvider) ?? "CDN" : string.Empty;
+                    _MyObservStreamingEndpoints[index].CDN = ((bool)streamingEndpoint.Data.IsCdnEnabled.GetValueOrDefault()) ? StreamingEndpointInformation.ReturnDisplayedProvider(streamingEndpoint.Data.CdnProvider) ?? "CDN" : string.Empty;
```
Comment (dima-ravnur) [inline AMSExplorer/DataGrid/DataGridViewStreamingEndpoints.cs]: same here

File: AMSExplorer/DataGrid/DataGridViewStreamingEndpoints.cs [hunk 4]
```diff
@@ -181,7 +181,7 @@ await foreach (var se in streamingendpointsPage)
                                 Name = c.Data.Name,
                                 Id = c.Data.Id,
                                 Description = c.Data.Description,
-                                CDN = (bool)c.Data.IsCdnEnabled ? StreamingEndpointInformation.ReturnDisplayedProvider(c.Data.CdnProvider) ?? "CDN" : string.Empty,
+                                CDN = c.Data.IsCdnEnabled.GetValueOrDefault() ? StreamingEndpointInformation.ReturnDisplayedProvider(c.Data.CdnProvider) ?? "CDN" : string.Empty,
```
Comment (dima-ravnur) [inline AMSExplorer/DataGrid/DataGridViewStreamingEndpoints.cs]: and here

File: AMSExplorer/Forms-Login/AMSLogin.cs [hunk 5]
```diff
@@ -234,11 +240,12 @@ private void ButtonExport_Click(object sender, EventArgs e)
                 exportSPSecrets = form.checkBoxIncludeSPSecrets.Checked;
 
                 PropertyRenameAndIgnoreSerializerContractResolver jsonResolver = new();
-                List<string> properties = new() { "EncryptedADSPClientSecret" };
+                List<string> properties = new() { "EncryptedADSPClientSecret", "MKIOEncryptedToken", "RavnurEncryptedApiKey" };
```
Comment (dima-ravnur) [inline AMSExplorer/Forms-Login/AMSLogin.cs:237]: Why MKIOEncryptedToken was missing here? Doesn't it work without it?

File: AMSExplorer/AMSClient/AMSClientV3.cs [hunk 6]
```diff
@@ -194,20 +200,29 @@ public async Task<MediaServicesAccountResource> ConnectAndGetNewClientV3Async(Fo
             credentials = new TokenCredentials(authResult.AccessToken, "Bearer");
             credentialForArmClient = new BearerTokenCredential(authResult.AccessToken);
 
+            //var credential = new DefaultAzureCredential(includeInteractiveCredentials: true);
+            ArmClient armClient = null;
+
+            if (IsRavnurClient)
+            {
+                armClient = RavnurClientFactory.GetRavnurArmClient(credentialsEntry);
```
Comment (dima-ravnur) [inline AMSExplorer/AMSClient/AMSClientV3.cs:207]: Lets add some comment here that it is a key point where we make AMS SDK work with RMS API

File: AMSExplorer/Mainform.cs [hunk 7]
```diff
@@ -281,30 +287,72 @@ public Mainform(string[] args)
                     TextBoxLogWriteLine("There is no streaming endpoint running in this account.", true); // Warning
                 }
 
-                var leResults = _amsClient.AMSclient.GetMediaLiveEvents().GetAllAsync().ToListAsync().Result;
-                double nbLiveEvents = leResults.Count();
```
Comment (dima-ravnur) [inline AMSExplorer/Mainform.cs:285]: Don't we break existing functionality for AMS?

File: AMSExplorer/Mainform.cs [hunk 8]
```diff
@@ -7011,7 +7077,7 @@ private async Task ManageMenuOptionsAzureCDNAsync(ToolStripMenuItem disableAzure
             {
                 var se = streamingendpoints.FirstOrDefault();
                 bool sestopped = (se.Data.ResourceState == StreamingEndpointResourceState.Stopped);
-                bool cdnenabled = (bool)se.Data.IsCdnEnabled;
+                bool cdnenabled = (bool)se.Data.IsCdnEnabled.GetValueOrDefault();
```
Comment (dima-ravnur) [inline AMSExplorer/Mainform.cs]: can we avoid it here as well?

**Review by dima-ravnur — CHANGES_REQUESTED** [review]

File: AMSExplorer/DataGrid/DataGridViewStreamingEndpoints.cs (hunk 2 above)
Comment (vstratovych) [reply AMSExplorer/DataGrid/DataGridViewStreamingEndpoints.cs]: Yes, the value of the field returned from our API is null.

This is totally acceptable according to the nullable value in the model, so this is actually issue in the AMS Explorer code.

However, I agree that we can make change on our side in order to simplify overall complexity of this PR.

File: AMSExplorer/Forms-Login/AMSLogin.cs (hunk 5 above)
Comment (vstratovych) [reply AMSExplorer/Forms-Login/AMSLogin.cs:237]: That seems to be a bug introduced when this property was added.

Later in the code there is a comment that describes intention not to export encrypted secrets.
Moreover, the property in CredentialsEntryV4 class contains comment that is very specific about ignoring that property during export.

File: AMSExplorer/Mainform.cs (hunk 7 above)
Comment (vstratovych) [reply AMSExplorer/Mainform.cs:285]: No, I don't think so.
We just made sure that live events API isn't called for RMS.
For AMS everything should work just as it was before.

File: AMSExplorer/Mainform.cs (hunk 8 above)
Comment (vstratovych) [reply AMSExplorer/Mainform.cs]: I'm going to revert this change and do fix in RMS API instead.

**Review by dima-ravnur — APPROVED** [review]

**Pull request merged**